todo_app/app.py

```diff
@@ -1,72 +1,131 @@
+    client_id=os.getenv('GITHUB_CLIENT_ID')
+    client_secret=os.getenv('GITHUB_SECRET')
+    app.secret_key=os.getenv('SECRET_KEY')
+
+    login_manager = LoginManager()
+    login_manager.init_app(app)
+
+    @login_manager.unauthorized_handler
+    def unauthenticated():
+        client = WebApplicationClient(client_id) 
+        uri = client.prepare_request_uri('https://github.com/login/oauth/authorize')
+        return redirect(uri)    
+        
+    @login_manager.user_loader
+    def load_user(user_id):
+        return User(user_id)
+    
+    @app.route('/login/callback', methods=['GET', 'POST'])
+    def callback():
+        code = request.values['code']
+        client = WebApplicationClient(client_id) 
+        url, headers, body = client.prepare_token_request('https://github.com/login/oauth/access_token', code=code)
+        git_access_key = requests.post(url, headers=headers, data=body, auth=(client_id, client_secret))
+        git_json = client.parse_request_body_response(git_access_key.text)
+        git_user_request = client.add_token("https://api.github.com/user")
+        git_user = requests.get(git_user_request[0], headers=git_user_request[1]).json()
+        git_login = User(git_user['login'])
+        login_user(git_login)
+        return redirect('/')
+
     @app.route('/')
+    @login_required
     def index():
         mongo = Mongo()
         to_do_list = mongo.get_todo_items() 
         yesterday = date.today() - timedelta(days = 1) 
 
-        to_do_view_model = ToDoViewModel(to_do_list, yesterday)
+        to_do_view_model = ToDoViewModel(to_do_list, yesterday, user_has_write_permissions())
         if (len(to_do_view_model.done_items) < 5):
             return render_template('index.html', view_model=to_do_view_model, show_all=True)
         else:
             return render_template('index.html', view_model=to_do_view_model, show_all=False)
             
     @app.route('/', methods=['POST'])
+    @login_required
     def add_to_do_item():
-        mongo = Mongo()
-        mongo.add_item(request.form['todoTitle'])
+        if (user_has_write_permissions() == False):
+            flash("You do not have permissions to do this action")
+        else:
+            mongo = Mongo()
+            mongo.add_item(request.form['todoTitle'])
         return redirect("/")
 
     @app.route('/doing_item/<card_id>')
+    @login_required
     def doing_item(card_id):
-        mongo = Mongo()
-        mongo.move_to_doing(card_id)
+        if (user_has_write_permissions() == False):
+            flash("You do not have permissions to do this action")
+        else:
+            mongo = Mongo()
+            mongo.move_to_doing(card_id)
         return redirect("/")
 
     @app.route('/todo_item/<card_id>')
+    @login_required
     def todo_item(card_id):
-        mongo = Mongo()
-        mongo.move_to_todo(card_id)
+        if (user_has_write_permissions() == False):
+            flash("You do not have permissions to do this action")
```

Comment:
You could try making a function decorator to avoid having to copy/paste this logic and message.